stdlib/public/runtime/SwiftObject.mm

```diff
@@ -373,11 +374,72 @@ + (BOOL)conformsToProtocol:(Protocol*)proto {
 }
 
 - (NSUInteger)hash {
+  auto selfMetadata = _swift_getClassOfAllocated(self);
+
+  // If it's Hashable, use that
+  auto hashableConformance =
+    reinterpret_cast<const hashable_support::HashableWitnessTable *>(
+      swift_conformsToProtocolCommon(
+	selfMetadata, &hashable_support::HashableProtocolDescriptor));
+  if (hashableConformance != NULL) {
+    return _swift_stdlib_Hashable_hashValue_indirect(
+      &self, selfMetadata, hashableConformance);
+  }
+
+  // If a type is Equatable (but not Hashable), we
+  // have to return something here that is compatible
+  // with the `isEqual:` below.  NSObject's default
+  // of `(NSUInteger)self` won't work, so we instead
+  // return a constant fallback value:
+  auto equatableConformance =
+    reinterpret_cast<const equatable_support::EquatableWitnessTable *>(
+      swift_conformsToProtocolCommon(
+	selfMetadata, &equatable_support::EquatableProtocolDescriptor));
+  if (equatableConformance != nullptr) {
+    return (NSUInteger)1;
+  }
+
+  // Legacy default for types that are neither Hashable nor Equatable.
   return (NSUInteger)self;
```

Comment:
@lorentey @mikeash I'm a little torn on this one.  I experimentally changed `-hash` to have three cases:
* Use the Swift Hashable conformance if present
* Return a constant if this is Swift Equatable but not Swift Hashable (this is necessary to be compatible with an arbitrary `==`)
* Return `self` if this is neither Swift Equatable nor Swift Hashable.

These are arguably the optimal hash values for each case, but there's a lot of dynamic overhead here.  Should I just go back to returning a constant for both of the last two cases?

Reply:
The overhead shouldn't be too bad. Looking up a cached conformance isn't as fast as, say, an `objc_msgSend`, but it's pretty quick. If this is somehow a performance issue for someone, they can pretty easily resolve it by implementing their own `hash` and `isEqual:`.

Reply:
So it sounds like the three-way split here seems the best option in terms of semantics.   I agree that a constant hash is a bad idea, but I don't see any other really viable option for this specific case.  I could however add a log message here calling out that someone tried to hash a Swift type that isn't Hashable and that this can lead to performance problems.  (Trapping is tempting, but that feels like a Swift answer, not an Obj-C answer.  🤔)

> ... we'll want __SwiftValue to implement exactly the same logic ...
Of course.  I have those changes locally and will submit them as a separate PR once we've nailed down the behavior we want here.

Reply:
Logging a runtime message is a really good idea! Do we have a rate-limited logging facility?

(We'd want to use it from Swift parts of the stdlib too.)

Reply:
Added.

Reply:
I don't think we have one already. Rather than rate-limiting, it might be good to limit it to once per class. We'd have to keep a set of classes we've logged for, but if the whole point is to call out performance problems, maybe that's OK.

Reply:
Any good examples of handling such a set (thread-safety, etc)?

Reply:
Unchecked continuation validation is pretty much this. Only difference is that it also removes items, but you don't need that part. Link: https://github.com/apple/swift/blob/c1e892418fff4a91d9abf5e957b21f748fd28e7a/stdlib/public/Concurrency/Task.cpp#L1280

Reply:
I went ahead with a std::unordered_set for this, copying the example from continuation validation.  I also updated the test to verify that the runtime does in fact emit one message for each type that hits this particular bit of code.

Reply:
Seems fine to use `unordered_set`. I don't have a good sense of the tradeoffs between the two but `unordered_set` clearly does work. We can always switch it if trouble arises.